libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+        }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="storageAccount">Azure CloudStorageAccount instance</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null) throw new ArgumentNullException(nameof(storageAccount));
+
+            // Checks if a container name is valid
+            NameValidator.ValidateContainerName(containerName);
+
+            this.Container = new Lazy<CloudBlobContainer>(() =>
+              {
+                  var blobClient = storageAccount.CreateCloudBlobClient();
+                  var container = blobClient.GetContainerReference(containerName);
+                  container.CreateIfNotExistsAsync().Wait();
+                  return container;
+              });
+        }
+
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
+            {
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
+                await blobReference.DeleteAsync(
+                    DeleteSnapshotsOption.None, new AccessCondition { IfMatchETag = "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve entities from the configured blob container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task<StoreItems> Read(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            var storeItems = new StoreItems();
+            foreach (string blobName in keys.Select(GetBlobName))
+            {
+                using (var memoryStream = new MemoryStream())

```

Comment:
If we don't change this implementation to process in parallel, we should allocate the `MemoryStream` outside the loop and reuse it because it makes no sense to allocate more than one and constantly regrow the buffer, that just creates a lot of GC pressure.

Now, if we _do_ parallelize we can't share the `MemoryStream` (simply) and I wouldn't address that issue right now. At some point we should do another pass over the code base any place that we're using `MemoryStreams` and consider using a memory pool such as .NET Core's built in `ArrayPool<T>` and wrapping those with `MemoryStream`.